src/basicfuns.jl

```diff
@@ -152,9 +152,50 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(x::Real)
+    x0, x1, x2 = _log1pexp_thresholds(x)
+    if x ≤ x0
+        return exp(x)
+    elseif x ≤ x1
+        return log1p(exp(x))
+    elseif x ≤ x2
+        return x + exp(-x)
+    else
+        return oftype(x0, x)
+    end
+end
+
+#=
+returns a thresholds x0, x1, x2 such that:
+    * log1pexp(x) ≈ exp(x) for x ≤ x0
+    * log1pexp(x) ≈ log1p(exp(x)) for x0 < x ≤ x1
+    * log1pexp(x) ≈ x + exp(-x) for x1 < x ≤ x2
+    * log1pexp(x) ≈ x for x > x2
+=#
+@generated function _log1pexp_thresholds(::T) where {T<:Real}
```

Comment:
I am generally sympathetic to this approach, but would perhaps restrict to `T <: AbstractFloat`, to explicitly avoid types like `ForwardDiff.Dual`, etc, which are `<:Real`.

Reply:
Why avoid `ForwardDiff.Dual`? Note that `eps(ForwardDiff.Dual)` works fine here.

```
julia> eps(ForwardDiff.Dual(1.0,1.0))
2.220446049250313e-16
```